oeps/oep-0041-arch-async-server-event-messaging.rst

```diff
@@ -0,0 +1,463 @@
+clearer for listeners and don't requiring knowing implementation details. For
+instance, it is preferable to have a ``registration.completed`` event than to
+have a ``registration.updated`` event and leave it to clients to check the
+resulting message's payload for ``status: "complete"``.
+
+Actions should be past tense, to better align with our existing conventions
+around Django signals and learning analytics events (we're not completely
+consistent, but we tend towards past tense).
+
+Major Version
+~~~~~~~~~~~~~
+
+Example: ``v1``
+
+The last portion is the major version of the message, starting with ``v1``. All
+messages will have a major and minor version, with the minor version encoded in
+the ``minorversion`` attribute. New fields may be added to a message without
+incrementing the major version, but all additions *must* be backwards
+compatible. If you increment this to make a backwards-incompatible change, you
+will be expected to have a transition period where you are emitting the previous
+major version event as well as the new one.
+
+
+`time <https://github.com/cloudevents/spec/blob/master/spec.md#time>`_
+----------------------------------------------------------------------
+
+Example: ``"2020-02-23T09:00:00Z"``
+
+Timestamp that the event occurred, in UTC using `RFC 3339
+<https://tools.ietf.org/html/rfc3339>`_. If this event was sent because we
+created a new row in the database, we should pull this ``time`` directly from
+the ``created_at`` field in that model so that the output matches exactly. Do
+*not* call ``datetime.now()`` in these situations because we will get times
+that are just a few milliseconds offset from the database record of these
+actions. Do call ``datetime.now()`` if the event happens and has no
+corresponding database changes. If you are sending out multiple event messages
+describing the same occurance (e.g. a version 1 and version 2 of an event), they
+should have the *exact* same timestamp.
+
+
+Message Content Data Guidelines
+===============================
+
+These are general guidelines to consider when creating your messages. There can
+always be exceptional circumstances and use cases that require going against one
+of these guidelines, but try to default to these guidelines unless you're really
+sure about what you're doing.
+
+
+Events are Created by the Owning Subdomain
+------------------------------------------
+
+Teams at edX are broadly aligned to subdomains and roughly mapped to services.
+Services should not emit events for other subdomains. For instance, the
+ecommerce service is its own subdomain and should not be emitting ``catalog`` or
+``learning`` events. It is sometimes the case that a subdomain encompasses
+multiple services (e.g. Studio and Blockstore both operate on the
+``content_authoring`` subdomain).
+
+
+One Producer Service Per Event Type
+-----------------------------------
+
+Each event type should be emitted by one, and only one, service. That service is
```

Comment:
How will people be able to figure out if a given event name is already in use? Should events be namespaced to the owning service/sub-domains?

Reply:
@feanil had some thoughts around having either a centralized schema registry repo, or else having a separate one per business subdomain (e.g. "catalog", "learning"). That's all very fuzzy to me at this point.

Reply:
I'll put a small note in here to this effect before merging this tonight. Thank you!

Reply:
Yeah just seems like if we want events to be namespaced and have only one owner, we should either figure out a way to programmatically enforce it or at least give people a way to see if the namespace they want for their event is already taken.